### DISCUSSION TITLE
Table Interface

### DISCUSSION THREAD

**Discussion by ada-ada-ada-art** [directus/directus#5457] [Feature Requests]
It would be great to have an interface for the JSON data type, that allows for adding rows and columns in one interface.

While it's definitely a possibility to just create a new database table / Directus collection, this seems excessive, if you don't expect to reuse the columns and rows. 

In my mind, the table should:
1. Allow for a custom amount of rows & columns
2. Output a nested array of strings

In essence, it should just mimic a tiny spreadsheet. 

## Example usage
Running an online webshop with clothes requires showing size guides.
The most standard way of showing this is with tables.
The size guides are different for each product.

A sample size guide could look like this:
– | S | M | L | XL | XXL
------------ | ------------ | ------------- | ------------ | ------------- |------------
Chest | 94-98 | 99-102 | 103-106 | 107-112 | 113-118
Waist | 82-86 | 87-90 | 91-94 | 95-98 | 98-101

**Comment by rijkvanzanten** [comment]
> Output a nested array of strings

This is interesting! So the JSON output wouldn't contain the S/M/L/XL/XXL labels? I kinda assumed this would use something like

```json
[
  {
    "label": "Chest",
    "s": "94-98",
    "m": "99-102",
    "l": "103-106",
    "xl": "107-112",
    "xxl": "113-118"
  },
  {
    "label": "Waist",
    "s": "94-98",
    "m": "99-102",
    "l": "103-106",
    "xl": "107-112",
    "xxl": "113-118"
  }
]
```

for it's output, but am I correctly understanding that you're suggesting this?:

```json
[
  ["Chest", "94-98", "99-102", "103-106", "107-112", "113-118"],
  ["Waist", "94-98", "99-102", "103-106", "107-112", "113-118"]
]
```

**Reply by benhaynes** [reply]
* Output could be based on the field's datatype... `json` or `string` (Nested arrays... or even Markdown)
* Might be nice if the interface allowed sorting (clicking the column header) and saving the new order
* Is it important to know the difference between headers and normal rows?
* We should plan to append pseudo-elements on each row for "Delete"
* Need a clean way to add new rows/columns
* It will be hard to make this anything near full-featured... but we should look into the simple things like column widths, how we overflow, and column alignment, etc
* Interface Option: Default Column/Row Counts, Max Columns, Max Rows, Pre-Define Column Headers?

**Reply by ada-ada-ada-art** [reply]
Loving the enthusiasm about this! 💪 

@rijkvanzanten 
I think your "label suggestion" is a bit rigid. In my mind, the most "pure" representation of a simple table is a nested array.

So, to be specific, I was imagining something more like this:
```
[
  ["-", "S", "M", "L", "XL", "XXL"],
  ["Chest", "94-98", "99-102", "103-106", "107-112", "113-118"],
  ["Waist", "94-98", "99-102", "103-106", "107-112", "113-118"]
]
```

If I can get a reliable nested array, I can make workarounds for a loooot of stuff.

I can approximate it by using a repeater with a text field that receives comma-separated values. But it's reeeeeally open to human error. Biggest problem is if they miss a comma, which can mess up the amount of cells.

@benhaynes 
I don't think it's important whether it's a row or header. You can probably discern that some other way.
In a nested array, I'd just assume that the first one is a header.

All of those interface options would be really helpful! 
Min/Max columns and rows especially.

### Reference
For reference, this is how it looks in Contentful with the plugin we're using currently:
<img width="759" alt="Captura de Tela 2021-05-05 às 10 00 34" src="https://user-images.githubusercontent.com/7010555/117112062-ca87ff80-ad88-11eb-9140-2cf78ee97a72.png">

**Reply by rijkvanzanten** [reply]
Ohh, so the headers themselves aren't part of any configuration, but actually in the field value 🤔 Interesting!

**Reply by benhaynes** [reply]
Makes sense! In Contentful, how did you make those sizes bold? Did you set that as a header somewhere, or does it assume the first row is a header? I like the pure simplicity of just giving rows/columns and you do the rest. Seems like a pretty straightforward Interface to build if we find the right library. I have no issue with the nested array output, but also think this Interface should support the JSON type. 👍

**Reply by benhaynes** [reply]
Also, worth some thinking about a Display to pair with this. In other words, how would this field's value be displayed in the system? 

* `6 Columns / 5 Rows`?
* `30 Cells`?
* Hover for popover card with table? This only works for certain sizes.

**Reply by ada-ada-ada-art** [reply]
Yeah, I think it just assumes the first one is the header and thus boldifies it. 

So, purely based on our needs in Organic Basics, I think both `6 Columns / 5 Rows` and `30 cells` is helpful.
Might also be helpful if it lists how many of the cells are filled out. 

Perhaps `30 cells (6x5)`?

**Reply by benhaynes** [reply]
Makes sense, or reversing that in the Display:

* `6×5 Table: 24 Values`
* `6×5 Table (24 Values)`
* `6×5 Table with 24 Values`

Also, there could be an interface option for: `First Row Is Header` (boolean). As an aside, I wonder if `heading` or `header` is more appropriate? 🤔

**Reply by ada-ada-ada-art** [reply]
Isn't that the eternal question when dealing with headers / headings? 😅 

Looking at the HTML spec for tables, header seems to be used more often than heading:
https://www.w3.org/html/wiki/Elements/th
https://www.w3.org/html/wiki/Elements/thead

**Reply by ernst86913** [reply]
Sorting (drag and drop), deleting, adding and adding/inserting row would also be great... then there are many many use cases for this interface

**Reply by bcartier** [reply]
Has anyone started working on this as an extension?

**Reply by benhaynes** [reply]
The Core team hasn't, but it passes 80/20 and is something we'd be open to discussing.

**Comment by rijkvanzanten** [comment]
Heya! Thank you for taking the time to submit this request! 

It has been over **90 days**, and this discussion has not received at least **15 votes** from the community. This means that we don't feel like there's enough community interest to warrant further R&D into this topic at this time. :ice_cube:

This request will now be closed to keep our discussions tidy. Please reach out if you have any questions!

_For more information, see our [Feature Request Process](https://docs.directus.io/contributing/feature-request-process.html)._
